---
title: "ComBat-seq on simulated datasets - differential expression"
author: Yuqing Zhang
date: "`r Sys.Date()`"
output:
  html_notebook:
    theme: united
    toc: yes
---

## ComBat-seq model

We developed a negative binomial GLM for adjusting for batch effect in count data. Gene-wise linear model:

$$
\log(\mu_{gij}) = \alpha_g + X_j^T\beta_g + \gamma_{gi} + \log(N_j) \\
var(y_{gij}) = \mu_{gij} + \phi_{gi}\mu_{gij}^2
$$

* $y_{gij}$: Counts of gene $g$ in sample $j$ from batch $i$.
* $\mu_{gij}$: Expectation of counts of gene $g$ in sample $j$ from batch $i$.
* $\alpha_g$: Background expression of gene $g$.
* $X_j^T\beta_g$: Biological condition effect of sample $j$, design matrix and corresponding coefficients.
* $\gamma_{gi}$: Mean batch parameter for batch $i$ on gene $g$.
* $\phi_{gi}$: Dispersion batch parameter for batch $i$ on gene $g$.
* $N_j$: Library size (total number of counts) in sample $j$.

edgeR applies EB shrinkage on dispersion parameter $\phi_{gi}$, through weighted likelihood. But it doesn't apply shrinkage on $\gamma_{gi}$. This means that, in theory, ComBat-seq should out-perform edgeR (including batch as covariate) when ...



## Simulation setup

We test how well ComBat-seq performs, compared to using other methods to address batch effect, in differential expression analysis. Our primary study design is:

<img src="/Users/yuqingz/Dropbox/Work/ComBat_Seq/study_design_4.png" width="425">

In this design, we assume that batch effect is in the form of both **fold changes in the average of counts** and **differences in dispersion** across batches. We also simulate the count matrix using the *polyester* package, which is convenient because it uses the negative binomial model of edgeR. These are consistant with our ComBat-seq model assumptions. 

We perform DE with edgeR, and compare TPR and FPR in the following 3 settings:

+ original counts, not including batch as a covariate
+ original counts, with batch included as a covariate
+ counts adjusted by ComBat-seq

Several factors can affect the performance of DE detection:

* Fold change of counts caused by batch / condition (mean batch effect)
* Dispersion differences (variance batch effect)
* Balanced / unbalanced study design
* Number of samples in total, and in batch / condition groups
* Number of reads per gene (sequencing depth)

We systematically evaluate ComBat-seq in different combinations of these factors, to see whether / when ComBat-seq works better than including batch as a covariate in edgeR.

```{r, echo=FALSE, results='hide'}
sapply(c("knitr", "kableExtra", "ggplot2", "gridExtra", "reshape2"), require, character.only=TRUE)
results_dir <- "~/Google Drive/ComBat_seq/DE_analysis_powerdown_unbalanced"
```



## Systematic evaluation {.tabset}

Boxplots show distributions of TPR and FPR from 100 simulations.

Setup start point:

* Fold changes: Biological - 1.5, Batch - 1.5
* Dispersion: Batch 1 - 0.01, Batch 2 - 0.1
* Number of samples: 10 in total 
* Reads per gene: Median 120 (coverage 5)

Note that this setup was arbitrarily selected - based on my exploration of real count data, it seems a okay place to start. Based on this setup, we vary each factor in turn, while keeping all the other factors fixed.


### Balanced (No confounding) {.tabset}

### Fold change of counts caused by batch / condition 

Test fold change of counts at these levels:

```{r, echo=FALSE, results='hide'}
FC_df <- t(data.frame(Biological=rep("1.5",4), Batch=c("1", "1.5", "2", "3")))
colnames(FC_df) <- paste("Fold change", 1:4)
kable(FC_df) %>%
  kable_styling(bootstrap_options = "striped", full_width = F)
```

```{r, echo=FALSE, results='hide', message=FALSE, fig.width=4, fig.height=8}
batch_fold_seq <- c(1, 1.5, 2, 3)
p_tpr_lst <- p_fpr_lst <- list()

for(i in 1:length(batch_fold_seq)){
  tpr_file <- sprintf("tpr_simFC_bio15_batch%s_sizes100_10_N10_U_depth5.csv", gsub('.', '', batch_fold_seq[i], fixed=TRUE))
  fpr_file <- sprintf("fpr_simFC_bio15_batch%s_sizes100_10_N10_U_depth5.csv", gsub('.', '', batch_fold_seq[i], fixed=TRUE))

  tpr_res <- read.csv(file.path(results_dir, tpr_file))
  fpr_res <- read.csv(file.path(results_dir, fpr_file))
  
  p_tpr_lst[[i]] <- ggplot(melt(tpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    labs(x="", y="TPR", title=sprintf("Bio FC 1.5, Batch FC %s", batch_fold_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  p_fpr_lst[[i]] <- ggplot(melt(fpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    geom_hline(yintercept=0.05, col="red") +
    labs(x="", y="FPR", title=sprintf("Bio FC 1.5, Batch FC %s", batch_fold_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  
  rm(tpr_file, tpr_res, fpr_file, fpr_res)
}

grid.arrange(p_fpr_lst[[1]], p_tpr_lst[[1]], p_fpr_lst[[2]], p_tpr_lst[[2]],
             p_fpr_lst[[3]], p_tpr_lst[[3]], p_fpr_lst[[4]], p_tpr_lst[[4]], 
             ncol=2, widths=rep(2,2), heights=rep(2,4))
rm(p_fpr_lst, p_tpr_lst, batch_fold_seq)
```


### Dispersion differences 

Test dispersion differences across batch at these levels:

```{r, echo=FALSE, message=FALSE, results='hide'}
Disp_df <- t(data.frame(Batch1=c("0.1", "0.01", "0.001"), Batch2=rep("0.1", 3)))
colnames(Disp_df) <- paste("Dispersion", 1:3)
kable(Disp_df) %>%
  kable_styling(bootstrap_options = "striped", full_width = F)
```

```{r, echo=FALSE, message=FALSE, results='hide', fig.width=4, fig.height=6}
size1_seq <- c(10, 100, 1000)
p_tpr_lst <- p_fpr_lst <- list()

for(i in 1:length(size1_seq)){
  tpr_file <- sprintf("tpr_simDisp_bio15_batch15_sizes%s_10_N10_U_depth5.csv", size1_seq[i])
  fpr_file <- sprintf("fpr_simDisp_bio15_batch15_sizes%s_10_N10_U_depth5.csv", size1_seq[i])

  tpr_res <- read.csv(file.path(results_dir, tpr_file))
  fpr_res <- read.csv(file.path(results_dir, fpr_file))
  
  p_tpr_lst[[i]] <- ggplot(melt(tpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    labs(x="", y="TPR", title=sprintf("Batch1 Disp %s, Batch2 Disp 0.1", 1/size1_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  p_fpr_lst[[i]] <- ggplot(melt(fpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    geom_hline(yintercept=0.05, col="red") +
    labs(x="", y="FPR", title=sprintf("Batch1 Disp %s, Batch2 Disp 0.1", 1/size1_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  
  rm(tpr_file, tpr_res, fpr_file, fpr_res)
}

grid.arrange(p_fpr_lst[[1]], p_tpr_lst[[1]], p_fpr_lst[[2]], p_tpr_lst[[2]],
             p_fpr_lst[[3]], p_tpr_lst[[3]], #p_fpr_lst[[4]], p_tpr_lst[[4]], 
             ncol=2, widths=rep(2,2), heights=rep(2,3))
rm(p_fpr_lst, p_tpr_lst, size1_seq)
```


### Sample size

Test sample sizes at these levels:

```{r, echo=FALSE, results='hide'}
N_df <- t(data.frame(c("10", "20", "40")))#, "80")))
colnames(N_df) <- paste("Sample size", 1:3); rownames(N_df) <- ""
kable(N_df) %>%
  kable_styling(bootstrap_options = "striped", full_width = F)
```

```{r, echo=FALSE, message=FALSE, results='hide', fig.width=4, fig.height=6}
N_seq <- c(10, 20, 40)#, 80)
p_tpr_lst <- p_fpr_lst <- list()

for(i in 1:length(N_seq)){
  tpr_file <- sprintf("tpr_simNsample_bio15_batch15_sizes100_10_N%s_U_depth5.csv", N_seq[i])
  fpr_file <- sprintf("fpr_simNsample_bio15_batch15_sizes100_10_N%s_U_depth5.csv", N_seq[i])

  tpr_res <- read.csv(file.path(results_dir, tpr_file))
  fpr_res <- read.csv(file.path(results_dir, fpr_file))
  
  p_tpr_lst[[i]] <- ggplot(melt(tpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    labs(x="", y="TPR", title=sprintf("Total sample size = %s", N_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  p_fpr_lst[[i]] <- ggplot(melt(fpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    geom_hline(yintercept=0.05, col="red") +
    labs(x="", y="FPR", title=sprintf("Total sample size = %s", N_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  
  rm(tpr_file, tpr_res, fpr_file, fpr_res)
}

grid.arrange(p_fpr_lst[[1]], p_tpr_lst[[1]], p_fpr_lst[[2]], p_tpr_lst[[2]], p_fpr_lst[[3]], p_tpr_lst[[3]], 
             ncol=2, widths=rep(2,2), heights=rep(2,3))
rm(p_fpr_lst, p_tpr_lst, N_seq)
```



### Sequencing depth 

Test reads per gene at these levels:

```{r, echo=FALSE, message=FALSE, results='hide'}
depth_df <- t(data.frame(c("24", "120", "240", "480")))
colnames(depth_df) <- paste("Median reads per gene", 1:4); rownames(depth_df) <- ""
kable(depth_df) %>%
  kable_styling(bootstrap_options = "striped", full_width = F)
```

```{r, echo=FALSE, message=FALSE, results='hide', fig.width=4, fig.height=8}
depth_seq <- c(1, 5, 10, 20)
p_tpr_lst <- p_fpr_lst <- list()

for(i in 1:length(depth_seq)){
  tpr_file <- sprintf("tpr_simDepth_bio15_batch15_sizes100_10_N10_U_depth%s.csv", depth_seq[i])
  fpr_file <- sprintf("fpr_simDepth_bio15_batch15_sizes100_10_N10_U_depth%s.csv", depth_seq[i])

  tpr_res <- read.csv(file.path(results_dir, tpr_file))
  fpr_res <- read.csv(file.path(results_dir, fpr_file))
  
  p_tpr_lst[[i]] <- ggplot(melt(tpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    labs(x="", y="TPR", title=sprintf("Coverage %s", depth_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  p_fpr_lst[[i]] <- ggplot(melt(fpr_res), aes(x=variable, y=value)) +
    geom_boxplot() +
    geom_hline(yintercept=0.05, col="red") +
    labs(x="", y="FPR", title=sprintf("Coverage %s", depth_seq[i])) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
  
  rm(tpr_file, tpr_res, fpr_file, fpr_res)
}

grid.arrange(p_fpr_lst[[1]], p_tpr_lst[[1]], p_fpr_lst[[2]], p_tpr_lst[[2]],
             p_fpr_lst[[3]], p_tpr_lst[[3]], p_fpr_lst[[4]], p_tpr_lst[[4]], 
             ncol=2, widths=rep(2,2), heights=rep(2,4))
rm(p_fpr_lst, p_tpr_lst, depth_seq)
```


