---
title: "Application of ComBat-seq on simulated dataset"
subtitle: "Performance in differential expression (DE)"
author: Yuqing Zhang
date: October 1, 2018
output: html_notebook
---

## ComBat-seq model

We developed a negative binomial GLM for adjusting for batch effect in count data. Gene-wise linear model:

$$
\log(\mu_{gij}) = \alpha_g + X_j^T\beta_g + \gamma_{gi} + \log(N_j) \\
var(y_{gij}) = \mu_{gij} + \phi_{gi}\mu_{gij}^2
$$

* $y_{gij}$: Counts of gene $g$ in sample $j$ from batch $i$.
* $\mu_{gij}$: Expectation of counts of gene $g$ in sample $j$ from batch $i$.
* $\alpha_g$: Background expression of gene $g$.
* $X_j^T\beta_g$: Biological condition effect of sample $j$, design matrix and corresponding coefficients.
* $\gamma_{gi}$: Mean batch parameter for batch $i$ on gene $g$.
* $\phi_{gi}$: Dispersion batch parameter for batch $i$ on gene $g$.

edgeR applies EB shrinkage on dispersion parameter $\phi_{gi}$, through weighted likelihood. But it doesn't apply shrinkage on $\gamma_{gi}$. This means that in theory, ComBat-seq should out-perform edgeR (including batch as covariate) when ...


## Simulation setup

We test how well ComBat-seq performs, compared to using other methods to address batch effect, in differential expression analysis. Our primary study design is:

<img src="/Users/yuqingz/Dropbox/Work/ComBat_Seq/study_design_2.png" width="350">

In this design, we assume that batch effect is in the form of both **fold changes in the average of counts** and **differences in dispersion** across batches. We also simulate the count matrix using the *polyester* package, which is convenient because it uses the negative binomial model of edgeR. These are consistant with our ComBat-seq model assumptions. 

We perform DE with edgeR, and compare TPR and FPR in the following 3 settings:

+ original counts, not including batch as a covariate
+ original counts, with batch included as a covariate
+ counts adjusted by ComBat-seq

Several factors can affect the performance of DE detection:

* Fold change of counts caused by batch / condition (mean batch effect)
* Dispersion differences (variance batch effect)
* Balanced / unbalanced study design
* Number of samples in total, and in batch / condition groups
* Number of reads per gene (sequencing depth)

We systematically evaluate ComBat-seq in different combinations of these factors, to see whether / when ComBat-seq works better than including batch as a covariate in edgeR.


## An example where ComBat-seq "works"...

* Fold changes: Biological - 1.5, Batch - 2 
* Dispersion: Batch 1 - 0.1, Batch 2 - 0.001
* Study design: Unbalanced 
* Number of samples: 20 in total 
* Reads per gene: Median 24 (coverage 1)





## Systematic evaluation 

All boxplots show distributions of TPR and FPR from 100 simulations.

### Fold change of counts caused by batch / condition {.tabset}

With the other parameters fixed:

* Dispersion differences: Batch 1 - , Batch 2 -
* Study design: 
* Number of samples: 
* Reads per gene:

We test fold change of counts at these levels:





### Dispersion differences {.tabset}




### Study design {.tabset}

With the other parameters fixed:

* Fold changes: Biological -, Batch -
* Dispersion differences: Batch 1 - , Batch 2 -
* Number of samples: 
* Reads per gene:

#### Balanced


#### Unbalanced



### Number of samples in total, and in batch / condition groups {.tabset}


### Sequencing depth {.tabset}



## Thoughts & future work

* Explore real count matrix to see if ComBat-seq assumptions are true: true WRT fold change, but more importantly, dispersion differences across batch?
* Try DESeq too
* When the total number of genes is large, ComBat-seq will slow down dramatically (i.e., if we simulate the count matrix to contain 20000 rows instead of 2000, the function will take >20mins to run). Need to optimize the algorithm.
