---
title: "Application of ComBat-seq on simulated dataset"
subtitle: "add random sparsity (0 counts) to simulate single-cell seq datasets"
author: Yuqing Zhang
date: September 29, 2018
output: html_notebook
---

## Design of the study

The **goal** of this simulation is to see whether ComBat-seq can be used on counts with 0s, i.e. single-cell sequencing datasets.

We simulated a **(2000 gene $\times$ 20 sample)** count matrix from negative binomial distributions. The dataset contains **2** batches, with **10** samples in each batch, and **biological effect - a 2-fold change between 2 condition groups**. 

<img src="/Users/yuqingz/Dropbox/Work/ComBat_Seq/study_design_2.png" width="400">


```{r, echo=FALSE, results='hide'}
rm(list=ls())
sapply(c("ggplot2", "reshape2", "gridExtra", "dendextend"), require, character.only=TRUE)
source("../ComBat_seq.R"); source("../helper_seq.R")
set.seed(123)

## Simulate count matrix and batch factor
y1_ctrl <- matrix(rnbinom(50*5, mu=10, size=20), ncol=5); #mean(c(y1_ctrl)); var(c(y1_ctrl))
y1_case <- matrix(rnbinom(50*5, mu=20, size=20), ncol=5); #mean(c(y1_case)); var(c(y1_case))
y1_null <- matrix(rnbinom(950*10, mu=10, size=20), ncol=10)

y2_ctrl <- matrix(rnbinom(50*5, mu=20, size=100), ncol=5); #mean(c(y2_ctrl)); var(c(y2_ctrl))
y2_case <- matrix(rnbinom(50*5, mu=40, size=100), ncol=5); #mean(c(y2_case)); var(c(y2_case))
y2_null <- matrix(rnbinom(950*10, mu=20, size=100), ncol=10)

y3_ctrl <- matrix(rnbinom(50*5, mu=20, size=20), ncol=5); #mean(c(y1_ctrl)); var(c(y1_ctrl))
y3_case <- matrix(rnbinom(50*5, mu=40, size=20), ncol=5); #mean(c(y1_case)); var(c(y1_case))
y3_null <- matrix(rnbinom(950*10, mu=20, size=20), ncol=10)

y4_ctrl <- matrix(rnbinom(50*5, mu=10, size=100), ncol=5); #mean(c(y2_ctrl)); var(c(y2_ctrl))
y4_case <- matrix(rnbinom(50*5, mu=20, size=100), ncol=5); #mean(c(y2_case)); var(c(y2_case))
y4_null <- matrix(rnbinom(950*10, mu=10, size=100), ncol=10)

counts <- rbind(cbind(y1_ctrl, y1_case, y2_ctrl, y2_case),
                cbind(y1_null, y2_null),
                cbind(y3_ctrl, y3_case, y4_ctrl, y4_case),
                cbind(y3_null, y4_null))
rownames(counts) <- paste0("gene", 1:nrow(counts))
colnames(counts) <- paste0("sample", 1:ncol(counts))

batch <- c(rep("B", ncol(y1_ctrl)+ncol(y1_case)), 
           rep("A", ncol(y2_ctrl)+ncol(y2_case))); table(batch)
group <- c(rep(0, ncol(y1_ctrl)), rep(1, ncol(y1_case)),
           rep(0, ncol(y2_ctrl)), rep(1, ncol(y2_case)))
full_mod <- TRUE
```

```{r, echo=FALSE, results='hide'}
add0 <- function(p_zeros, counts, batch, group){
  # add zeros to count matrix
  #p_zeros <- 0.5  # percentage of zeros in the count matrix
  n_zeros <- round(nrow(counts)*ncol(counts)*p_zeros, 0)
  sel_locs <- sample(1:(nrow(counts)*ncol(counts)), n_zeros, replace=FALSE)
  sel_mat <- rep(FALSE, nrow(counts)*ncol(counts))
  sel_mat[sel_locs] <- TRUE
  sel_mat <- matrix(sel_mat, nrow=nrow(counts), ncol=ncol(counts))
  counts[sel_mat] <- 0

  # remove genes with all 0 counts
  zero_genes_ind <- apply(counts,1,function(x){all(x==0)})
  cat("Number of genes with 0 in all samples:\n")
  cat(sum(zero_genes_ind))
  cat("\n\n")
  counts <- counts[!zero_genes_ind, ]

  # apply ComBat-seq
  if(nrow(counts)!=0){
    invisible(capture.output(adj_counts <- ComBat_seq(counts=counts, batch=batch, group=group)))
  }else{
    adj_counts <- NULL
  }
  return(adj_counts)
}
```


## Distribution of adjusted counts in samples {.tabset}

### 0% zero counts added

```{r, echo=FALSE, results='hide'}
invisible(capture.output(adj_counts_0 <- ComBat_seq(counts=counts, batch=batch, group=group)))
```

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_0))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_0)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_0)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```

### 10% zero counts added

```{r, echo=FALSE}
adj_counts_1 <- add0(p_zeros=0.1, counts, batch, group)
```

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_1))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_1)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_1)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```




### 20% zero counts added

```{r, echo=FALSE}
adj_counts_2 <- add0(p_zeros=0.2, counts, batch, group)
```

#### Distribution of adjusted counts in samples

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_2))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_2)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_2)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```


### 30% zero counts added

```{r, echo=FALSE}
adj_counts_3 <- add0(p_zeros=0.3, counts, batch, group)
```

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_3))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_3)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_3)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```


### 40% zero counts added

```{r, echo=FALSE}
adj_counts_4 <- add0(p_zeros=0.4, counts, batch, group)
```

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_4))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_4)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_4)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```


### 50% zero counts added

```{r, echo=FALSE}
adj_counts_5 <- add0(p_zeros=0.5, counts, batch, group)
```

```{r, echo=FALSE}
adj_counts_df_mlt <- data.frame(Batch=NA, melt(adj_counts_5))
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_5)[batch=="A"]] <- "A"
adj_counts_df_mlt$Batch[adj_counts_df_mlt$Var2 %in% colnames(adj_counts_5)[batch=="B"]] <- "B"
adj_counts_df_mlt$Batch <- as.factor(adj_counts_df_mlt$Batch)

## Boxplot of the adjusted counts in each sample
ggplot(adj_counts_df_mlt, aes(x=Var2, y=value, fill=Batch)) + 
  geom_boxplot() +
  labs(title="Distribution of adjusted counts in each sample",x="Samples", y = "Counts") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank())
```


## Clustering of samples after adjustment {.tabset}

### 0% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_0)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```

### 10% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_1)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```

### 20% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_2)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```

### 30% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_3)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```

### 40% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_4)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```

### 50% zero counts added

```{r, echo=FALSE}
adj_hc <- hclust(dist(t(adj_counts_5)))
adj_dend <- as.dendrogram(adj_hc)
adj_dend <- color_branches(adj_dend, groupLabels=batch[order.dendrogram(adj_dend)], 
                           col=group[order.dendrogram(adj_dend)]+2)
plot(adj_dend)  
```


## Thoughts

* It seems that as the proportion of 0s increases, ComBat-seq tends to generate larger outlier counts. It handles batch effect worse, too.
* However, randomly assigning positions in count matrix to be 0 is not a good way to mimic single-cell sequencing data. Could consider scRNA-seq simulator (like Splatter).

